### PR TITLE
Fix GPU compilation for newer CUDA versions

### LINKED ISSUE #62: Improve GPU detection script
With the current version of GPU detection script, CMake configuration fails for new CUDA versions, which have not yet been added to the script.

While this is not a lot of work, it's not very pleasant for the user. For example, someone may be happy with an old version of Ginkgo, and wants to compile it with a newer CUDA version. This will of course sometimes be impossible if the new CUDA API has changed, but if it didn't there is no reason why it wouldn't compile.

Thus, removing this restriction from the detection script would be a good idea. It should be possible to do by extracting the supported architectures directly from `nvcc --help` command instead of relying on hardcoded values. We still won't get support for new generation names (`Fermi`, `Maxwell`, etc.), but at least the user will be able to compile Ginkgo by specifying the architectures explicitly.

### PR DESCRIPTION
**Pull request by gflegar**
This PR fixes the problem with the GPU detection script we encountered in #141 (and will affect all our CI builds until this PR is merged). This fixes #62.

The reason for the failures is that GPU detection was using a static list of supported GPU architectures for each CUDA version. Whenever a newer CUDA version appeared, the script would have to be updated with the list for the new CUDA version. Otherwise it was throwing an error as the list was not found, and was considered empty.

While fixing that issue I decided to completely rewrite the GPU detection module, resulting in the following benefits:

*   The new module is completely independent of Ginkgo-specific behavior, so it was extracted into a separate project (https://github.com/ginkgo-project/CudaArchitectureSelector), and included as a dependency to Ginkgo. This means that the same code can now be easily adopted by other projects.
*   The list of supported architectures is no longer hardcoded, but is instead obtained from the compiler's documentation. Unfortunately, I didn't find a way to obtain the information in machine readable form, so I have to parse the output of `nvcc --help`, which could potentially break in the future if NVIDIA decides to change the way the information is presented. But the script will work as long as they keep the format roughly the same, and just add new and remove obsolete architectures from the documentation.
*   The interface of the function was modified to follow [modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1) - i.e. setting properties to targets instead of using global variables (we are yet to fully move Ginkgo to this new paradigm)
*   Better documentation of the new module (embedded in its source). I did remove the "MaxPTX" and "Off" options though, since the former is useless ("MinPTX" would have more sense, but is impossible to do without project-specific knowledge) and the latter can be achieved by passing an empty list. Also the "Auto" option will no longer fall back to "All" (which doesn't make much sense since that's probably not something anyone wants to do - otherwise, the project could just specify "All" by default), but to nvcc's default (which is to generate PTX for the lowest supported virtual architecture).

The GPU tests still fail with this commit, since #141 is not yet merged (and I would like to merge this PR before #141, so we can successfully test it on the CI).

### CODE REVIEW

File: cuda/CMakeLists.txt [hunk 1]
```diff
@@ -18,17 +18,13 @@ set(SOURCES
     stop/residual_norm_reduction_kernels.cu)
 
 enable_language(CUDA)
-include(../cmake/detect_gpu.cmake)
+include(${CAS_SOURCE})
```
Comment (tcojean) [inline cuda/CMakeLists.txt]: I guess this comes from CAS? This feels a bit confusing due to the way CMake works.

File: README.md [hunk 2]
```diff
@@ -123,18 +123,14 @@ Ginkgo adds the following additional switches to control what is being built:
...
-    compiler. `Off` will not select any architectures and compile with NVCC's
-    default settings. Default is `Auto`.
+    *   `CODE`, `CODE(COMPUTE)`, `(COMPUTE)`
+
+    `Auto` will automatically detect the present CUDA-enabled GPU architectures
+    in the system. `Kepler`, `Maxwell`, `Pascal` and `Volta` will add flags for
+    all architectures of that particular NVIDIA GPU generation. `COMPUTE` and
+    `CODE` are placeholders that should be replaced with compute and code
+    numbers (e.g.  for `compute_70` and `sm_70` `COMPUTE` and `CODE` should be
+    replaced with `70`. Default is `Auto`.
```
Comment (tcojean) [inline README.md]: Maybe you could somewhere add a link to the CAS repository, also indicating the users if they want more details they can find it there.

**Comment by tcojean** [comment]
Also add a mention that this closes #62.

File: cuda/CMakeLists.txt (hunk 1 above)
Comment (gflegar) [reply cuda/CMakeLists.txt]: From CAS's CMakeLists.txt, which is automatically added to the project by `load_git_repository`.

I could do an alternative: update the `CMAKE_MODULE_PATH` in CAS's `CMakeLists.txt`, but then I also need to propagate the module path to the parent in:

1. third_party/CudaArchitectureSelector/CMakeLists.txt
2. third_party/CMakeLists.txt

Then, this inlcude can be replaced with `include(CudaArchitectureSelector)`.

It is more code, but maybe it's easier to understand and a bit more standard way of doing things.

**Comment by gflegar** [comment]
> ```
> # All available architectures:
> #   Specified by the string `All``. This option will add flags for CUBIN code
> #   generation for all supported GPU architectures.
> ```

@tcojean do you think it would be clearer if it sad "... for all GPU architectures supported by the compiler."?

**Comment by tcojean** [comment]
> do you think it would be clearer if it sad "... for all GPU architectures supported by the compiler."?

@gflegar That sounds good!

File: cuda/CMakeLists.txt (hunk 1 above)
Comment (tcojean) [reply cuda/CMakeLists.txt]: That is maybe a good idea actually, as it is essentially a CMake module and this feels like the standard way of manipulating them.

**Comment by gflegar** [comment]
@tcojean I think I addressed all your comments, you can look into the code again if you want to

**Comment by tcojean** [comment]
I just did, I approve your changes. Thanks for this code!

**Pull request merged**